### ISSUE TITLE
-llibprocesslib missing in Lima.pc

### ISSUE THREAD

**Issue by picca** [esrf-bliss/Lima#31]
Hello, 

if I do an ldd on my machine, I get
$ ldd /usr/lib/liblimacore.so
        linux-gate.so.1 =>  (0xb7728000)
        libprocesslib.so.1.3 => /usr/lib/libprocesslib.so.1.3 (0xb743f000)
        librt.so.1 => /lib/i386-linux-gnu/i686/cmov/librt.so.1 (0xb7436000)
        libnexuscpp.so.0 => /usr/lib/i386-linux-gnu/libnexuscpp.so.0 (0xb73d8000)
        libyat.so.0 => /usr/lib/i386-linux-gnu/libyat.so.0 (0xb7336000)
        libNeXus.so.0 => /usr/lib/libNeXus.so.0 (0xb7318000)
        libmxml.so.1 => /usr/lib/libmxml.so.1 (0xb7310000)
        libhdf5_serial.so.8 => /usr/lib/i386-linux-gnu/libhdf5_serial.so.8 (0xb703f000)
        libmfhdf.so.0 => /usr/lib/libmfhdf.so.0 (0xb7012000)
        libtiff.so.4 => /usr/lib/i386-linux-gnu/libtiff.so.4 (0xb6fac000)
        libconfig++.so.9 => /usr/lib/libconfig++.so.9 (0xb6f96000)
        libstdc++.so.6 => /usr/lib/i386-linux-gnu/libstdc++.so.6 (0xb6eaa000)
        libm.so.6 => /lib/i386-linux-gnu/i686/cmov/libm.so.6 (0xb6e84000)
        libgcc_s.so.1 => /lib/i386-linux-gnu/libgcc_s.so.1 (0xb6e66000)
        libc.so.6 => /lib/i386-linux-gnu/i686/cmov/libc.so.6 (0xb6d02000)
        libgsl.so.0 => /usr/lib/libgsl.so.0 (0xb6abf000)
        libgslcblas.so.0 => /usr/lib/libgslcblas.so.0 (0xb6a7f000)
        libpthread.so.0 => /lib/i386-linux-gnu/i686/cmov/libpthread.so.0 (0xb6a66000)
        /lib/ld-linux.so.2 (0xb7729000)
        libxml2.so.2 => /usr/lib/i386-linux-gnu/libxml2.so.2 (0xb6911000)
        libhdf5.so.7 => /usr/lib/libhdf5.so.7 (0xb6630000)
        libdf.so.0 => /usr/lib/libdf.so.0 (0xb6573000)
        libjpeg.so.8 => /usr/lib/i386-linux-gnu/libjpeg.so.8 (0xb653a000)
        libz.so.1 => /lib/i386-linux-gnu/libz.so.1 (0xb6521000)
        libdl.so.2 => /lib/i386-linux-gnu/i686/cmov/libdl.so.2 (0xb651c000)
        libgfortran.so.3 => /usr/lib/i386-linux-gnu/libgfortran.so.3 (0xb6422000)
        libquadmath.so.0 => /usr/lib/i386-linux-gnu/libquadmath.so.0 (0xb63ab000)
        libjbig.so.0 => /usr/lib/i386-linux-gnu/libjbig.so.0 (0xb639d000)
        liblzma.so.5 => /lib/i386-linux-gnu/liblzma.so.5 (0xb6375000)

When I try to link something with limacore using Lima.pc

I got this error message

make[1]: entrant dans le répertoire « /tmp/Lima-Tango/LimaDetector/limadetector-1.5.2/_build »
g++ -I../src -I../src -I../specifics/SimulatorCCD -I../specifics/VieworksVP -std=c++11 -D__x86__ -D__linux__ -D__OSVERSION__=2 -I/usr/include/yat4tango/src -I/usr/include/yat4tango/include -I/usr/include/yat/src -I/usr/include/yat/include -I/usr/include/tango   -I/usr/include/yat/src -I/usr/include/yat/include   -I/usr/include/hdf5/serial   -DWITH_SPS_IMAGE -DWITH_NXS_SAVING -DWITH_TIFF_SAVING -DWITH_CONFIG -I/usr/include/Lima   -std=c++11 -D__x86__ -D__linux__ -D__OSVERSION__=2 -I/usr/include/tango   -I/opt/siso/include   -g -O2 -lyat4tango -lyat -ltango -lomniDynamic4 -lCOS4 -llog4tango -lzmq -lomniORB4 -lomnithread   -lyat   -L/usr/lib/i386-linux-gnu/hdf5/serial -lhdf5   -llimacore   -llimasimulator -llimavieworks-vp -ltango -lomniDynamic4 -lCOS4 -llog4tango -lzmq -lomniORB4 -lomnithread   -L/opt/siso/lib -lfglib5    -o ds_LimaDetector specifics/SimulatorCCD/SimulatorCCDClass.o specifics/SimulatorCCD/SimulatorCCDStateMachine.o specifics/SimulatorCCD/SimulatorCCD.o src/LimaDetectorClass.o src/LimaDetector.o src/LimaDetectorStateMachine.o src/main.o src/Factory.o src/ClassFactory.o src/AcquisitionTask.o specifics/VieworksVP/VieworksVP.o specifics/VieworksVP/VieworksVPStateMachine.o specifics/VieworksVP/VieworksVPClass.o  
/usr/bin/ld: /usr/lib/gcc/i486-linux-gnu/4.7/../../../../lib/liblimacore.so: undefined reference to symbol '_ZN4Data15HeaderContainerD1Ev'
/usr/bin/ld: note: '_ZN4Data15HeaderContainerD1Ev' is defined in DSO /usr/lib/libprocesslib.so.1.3 so try adding it to the linker command line
/usr/lib/libprocesslib.so.1.3: could not read symbols: Invalid operation

Will it be possible to add -lprocesslib in the Lima.pc.in file ?

thanks

Frederic

**Comment by esrf-bliss** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

Hello Fred,

Normally libprocesslib should not be in Lima.pc because liblimacore is
already link with it. All limacore's dependencies should be get  by
the compiler automatically.
It must be an other problem.
Is the processlib under /usr/lib is stripped?

Cheers,

SEB
On 06/05/2015 06:59 PM, picca wrote:
> Hello,
> 
> if I do an ldd on my machine, I get $ ldd /usr/lib/liblimacore.so 
> linux-gate.so.1 =>  (0xb7728000) libprocesslib.so.1.3 =>
> /usr/lib/libprocesslib.so.1.3 (0xb743f000) librt.so.1 =>
> /lib/i386-linux-gnu/i686/cmov/librt.so.1 (0xb7436000) 
> libnexuscpp.so.0 => /usr/lib/i386-linux-gnu/libnexuscpp.so.0
> (0xb73d8000) libyat.so.0 => /usr/lib/i386-linux-gnu/libyat.so.0
> (0xb7336000) libNeXus.so.0 => /usr/lib/libNeXus.so.0 (0xb7318000) 
> libmxml.so.1 => /usr/lib/libmxml.so.1 (0xb7310000) 
> libhdf5_serial.so.8 => /usr/lib/i386-linux-gnu/libhdf5_serial.so.8
> (0xb703f000) libmfhdf.so.0 => /usr/lib/libmfhdf.so.0 (0xb7012000) 
> libtiff.so.4 => /usr/lib/i386-linux-gnu/libtiff.so.4 (0xb6fac000) 
> libconfig++.so.9 => /usr/lib/libconfig++.so.9 (0xb6f96000) 
> libstdc++.so.6 => /usr/lib/i386-linux-gnu/libstdc++.so.6
> (0xb6eaa000) libm.so.6 => /lib/i386-linux-gnu/i686/cmov/libm.so.6
> (0xb6e84000) libgcc_s.so.1 => /lib/i386-linux-gnu/libgcc_s.so.1
> (0xb6e66000) libc.so.6 => /lib/i386-linux-gnu/i686/cmov/libc.so.6
> (0xb6d02000) libgsl.so.0 => /usr/lib/libgsl.so.0 (0xb6abf000) 
> libgslcblas.so.0 => /usr/lib/libgslcblas.so.0 (0xb6a7f000) 
> libpthread.so.0 => /lib/i386-linux-gnu/i686/cmov/libpthread.so.0
> (0xb6a66000) /lib/ld-linux.so.2 (0xb7729000) libxml2.so.2 =>
> /usr/lib/i386-linux-gnu/libxml2.so.2 (0xb6911000) libhdf5.so.7 =>
> /usr/lib/libhdf5.so.7 (0xb6630000) libdf.so.0 =>
> /usr/lib/libdf.so.0 (0xb6573000) libjpeg.so.8 =>
> /usr/lib/i386-linux-gnu/libjpeg.so.8 (0xb653a000) libz.so.1 =>
> /lib/i386-linux-gnu/libz.so.1 (0xb6521000) libdl.so.2 =>
> /lib/i386-linux-gnu/i686/cmov/libdl.so.2 (0xb651c000) 
> libgfortran.so.3 => /usr/lib/i386-linux-gnu/libgfortran.so.3
> (0xb6422000) libquadmath.so.0 =>
> /usr/lib/i386-linux-gnu/libquadmath.so.0 (0xb63ab000) libjbig.so.0
> => /usr/lib/i386-linux-gnu/libjbig.so.0 (0xb639d000) liblzma.so.5
> => /lib/i386-linux-gnu/liblzma.so.5 (0xb6375000)
> 
> When I try to link something with limacore using Lima.pc
> 
> I got this error message
> 
> make[1]: entrant dans le répertoire «
> /tmp/Lima-Tango/LimaDetector/limadetector-1.5.2/_build » g++
> -I../src -I../src -I../specifics/SimulatorCCD
> -I../specifics/VieworksVP -std=c++11 -D__x86__ -D__linux__
> -D__OSVERSION__=2 -I/usr/include/yat4tango/src
> -I/usr/include/yat4tango/include -I/usr/include/yat/src
> -I/usr/include/yat/include -I/usr/include/tango
> -I/usr/include/yat/src -I/usr/include/yat/include
> -I/usr/include/hdf5/serial   -DWITH_SPS_IMAGE -DWITH_NXS_SAVING
> -DWITH_TIFF_SAVING -DWITH_CONFIG -I/usr/include/Lima   -std=c++11
> -D__x86__ -D__linux__ -D__OSVERSION__=2 -I/usr/include/tango
> -I/opt/siso/include   -g -O2 -lyat4tango -lyat -ltango
> -lomniDynamic4 -lCOS4 -llog4tango -lzmq -lomniORB4 -lomnithread
> -lyat   -L/usr/lib/i386-linux-gnu/hdf5/serial -lhdf5   -llimacore
> -llimasimulator -llimavieworks-vp -ltango -lomniDynamic4 -lCOS4
> -llog4tango -lzmq -lomniORB4 -lomnithread   -L/opt/siso/lib
> -lfglib5    -o ds_LimaDetector
> specifics/SimulatorCCD/SimulatorCCDClass.o
> specifics/SimulatorCCD/SimulatorCCDStateMachine.o
> specifics/SimulatorCCD/SimulatorCCD.o src/LimaDetectorClass.o
> src/LimaDetector.o src/LimaDetectorStateMachine.o src/main.o
> src/Factory.o src/ClassFactory.o src/AcquisitionTask.o
> specifics/VieworksVP/VieworksVP.o
> specifics/VieworksVP/VieworksVPStateMachine.o
> specifics/VieworksVP/VieworksVPClass.o /usr/bin/ld:
> /usr/lib/gcc/i486-linux-gnu/4.7/../../../../lib/liblimacore.so:
> undefined reference to symbol '_ZN4Data15HeaderContainerD1Ev' 
> /usr/bin/ld: note: '_ZN4Data15HeaderContainerD1Ev' is defined in
> DSO /usr/lib/libprocesslib.so.1.3 so try adding it to the linker
> command line /usr/lib/libprocesslib.so.1.3: could not read symbols:
> Invalid operation
> 
> Will it be possible to add -lprocesslib in the Lima.pc.in file ?
> 
> thanks
> 
> Frederic
> 
> 
> --- Reply to this email directly or view it on GitHub: 
> https://github.com/esrf-bliss/Lima/issues/31
> 

-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2.0.22 (GNU/Linux)

iQIcBAEBAgAGBQJVdVLZAAoJEFf8JcAaf+jxLx4P/RloX57PqIPI3c+41+xvSVTk
8S6szQHGuMzJRuewABMNnTvPJif1E93ScGgcq+HC5xDkYAPVR6w9ah0r67OV1DMw
ioe+bXDMYXAT8cE4Fy3GWoM7mgQPSkZ7AkQkiIXoh/zx2oPkrAnLsPzWwAn2eMhg
ODSzKc0/qII+jtbVaONj2gjlHeqfYMggDLJ3qzCG6PL4uovgFBTaGcI+qqTrei1+
4mWj/CcDyDidrS8JEqWmBQoz6vTCCJvF0wF02w6yc+ddhIEiffN6Nbymx3qvb2l4
wyGpq+D/069knF/kIwrAw4QFdAM5uBUSsvxnybkv/k8HeARKMKGSHvgfZIQKjCYd
0wOWAyI4/Ly4qZjFQbR3KFUr0FoT0Z9Eiz1Ck8m9RbTIIRwVJqS+OD17KUsTNQ0i
jewqNbH4oRZRqbXkWogaZxOZWwS5j3/G76q9gc+xAuyNcG7DKLzDNKcwdXVPVKVg
GoDrWZnfkErB/HksCgJLzWm6v2xXAk49RPNOyN/BHaQpSlJN+KGeXaHybP4yMKtO
B2U7HI4nRstotnaf7hOBA8TTlELkQ5mNe8VSZipygXI31bGeFEfGOFBmfKJwAu+v
IGExr7l1Fb4mIRIdh5imdYMzxjWYf0xKezS96kFywxSqfAQdn2k3bSkzwf+yfqsw
as30IrjdiRwy6fwNN8wd
=iqHD
-----END PGP SIGNATURE-----

**Comment by picca** [comment]
yes I built a debian package so during the process, all the libraries are stripped and since rpath are not allowed, I removed all the rpath from the libraries.

**Comment by picca** [comment]
Hello, here the pkg-config guide which explin what should be put in the different field.
so it seems that if a library expose via it's public headers a thir party headers, this library should be added to the public part
cflags and libs.

depending on the built cameras, these fileds should be augmented with the camera related specific libraries.
http://people.freedesktop.org/~dbn/pkg-config-guide.html


I do not know what is your point of view about this Lima.pc
should it contain only the libraries used to build a camera or should it contain all the available libraries. ?

maybe a .pc files could be generated for each camera and create a Lima-camera-all.pc
Lima-camera-xxx.pc
...

**Comment by esrf-bliss** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

Hi Fred,

I don't know if it'll fixed the problem, but on new gcc order of
shared library matter. So could you put all shared libraries at the
end of your compilation line....

Just my two cents,

SEB
On 06/08/2015 11:16 AM, picca wrote:
> yes I built a debian package so during the process, all the
> libraries are stripped and since rpath are not allowed, I removed
> all the rpath from the libraries.
> 
> 
> 
> --- Reply to this email directly or view it on GitHub: 
> https://github.com/esrf-bliss/Lima/issues/31#issuecomment-109921841
>
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2.0.22 (GNU/Linux)

iQIcBAEBAgAGBQJVdW+/AAoJEFf8JcAaf+jxIe4P/3IysLIOjXSUTIrAwt99Onqq
K1oQhsae17/fOVRQ49vBiUJe7YIhNVs9kb92rXW3yxrtRClWNvvZicCEbIqT9hzl
fw39r48gPFKp7MEMaMT1/iVOvEtYWkw+nZCZUYMZbWr/9Qpd6nPtNbM1kOKVQ7BF
Cg83Ye1c5G29J6l0DpSWoMtf5gcqVnGN9dTpDzjVVfW8XKK4WsatIsDMzmjHawCY
FIlpfnqQzeJDxr+ioW8JdRBX14Wc5EM1QVgOUZWCJsuU2mo/t0ak+tsTs4DWhJVN
uz/B7p7BHQsX/PpTywPwxgtwGKv3Z+wvJ4esbC15pQ0g5UGuE5/wIg+Y3qYofJ4P
a/luRJuVeKecpS4KHbJXl5uOkosEEIHKfk39nDyn8yWuPQ80hIg9CA/UtNMowDGP
qQla/uToZjPE1zW2Rp3Gdw93Z4toOFSrt+ViWIJGHyw8iFusZ7pZI52ynotBPYdu
l61FBKDJwy5D0B9V0v0EqURzv6M599s9V/MuwOO4sPhjfDuEAefzT6k8/egU+RcF
NGIFjxgzPWH7EJd15Sv8qcjbrIt5ZvE0HezaipGTt+xnfa/nFixArHRHeTGYQdcA
ydefRXEv796l0UaeSaoDD8gHQ3qK9ZtxwiMx/S6BRCF1nDTv5VWfSTfComRNQZ8f
lohHZrsgIMRVTuqxUDOp
=NqGt
-----END PGP SIGNATURE-----

**Comment by picca** [comment]
I can not test now, but just by adding -lprocesslib it solved my problem :)

**Comment by esrf-bliss** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

Hi Fred,

I've checked some .pc files and most of them split dynamic and static
libraries with Libs and Libs.private

i.e:
	pkg-config --libs python
	pkg-config --static --libs python

As Lima make files don't manage static compilation, the file Lima.pc
seams correct to me. Am I wrong?

I saw that tango.pc export all dependencies even for dynamic
compilation, there is not difference between:
	pkg-config --libs tango
and
	pkg-config --static --libs tango

It's not really an issue but...

I think it's a good idea to generate one .pc file for each camera, we
will probably do that when we will move to autoconf (Lima 2.X)

We definitely have to move to a standard compilation tools like
autoconf and automake to get things right (.pc file,Makefile...). If
you (or other) have ideas about this subject, they are more then welcome
.

Regards,

SEB
in libraries even if they are private (tangolog,zmq...)
On 06/08/2015 11:34 AM, picca wrote:
> Hello, here the pkg-config guide which explin what should be put in
> the different field. so it seems that if a library expose via it's
> public headers a thir party headers, this library should be added
> to the public part cflags and libs.
> 
> depending on the built cameras, these fileds should be augmented
> with the camera related specific libraries. 
> http://people.freedesktop.org/~dbn/pkg-config-guide.html
> 
> 
> I do not know what is your point of view about this Lima.pc should
> it contain only the libraries used to build a camera or should it
> contain all the available libraries. ?
> 
> maybe a .pc files could be generated for each camera and create a
> Lima-camera-all.pc Lima-camera-xxx.pc ...
> 
> --- Reply to this email directly or view it on GitHub: 
> https://github.com/esrf-bliss/Lima/issues/31#issuecomment-109929559
>
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2.0.22 (GNU/Linux)

iQIcBAEBAgAGBQJVeWb6AAoJEFf8JcAaf+jxHtAQALLn4e3v2pdLVnqT7WMYitj/
6Odl5wMevrv9hhCh1n74gvKFzH2h1p/IbRfzEXMsxl+kEjAei0ocD6hZsMLpUE5/
PcmlIwCVZ5hCD2u/SLAxb4vPQzHWWl/P/IBA4oPPEn8jQCSRBGBKGr4y6YjqCWdg
7uZLMVfkqvNcOuKbMR/iaqC0Dya3nek941nnW1qaWkY8K+j7VZ3G4X2EhnHujfnk
zvqFMRtUu/7QwLBJ6O8qzD7l2GJNKCkC3OH3KvEly7QGTlzWffuit3FFRyRaNCZE
sOoZsdeNtwgvNDNa+431id/bdX20US7vIcQ3qPf3TN/coRpfd01Fkh/zfsPtp7oi
x1BxQR8NkE+Wr70FLT7yx5wctgCvDdybbAtugxbUVQUEZK6eLr4v6VALav/dtF2t
aapcWe60CmfEhMPMumsb5L1xApyUFKgKl24F1SC9VkXu0QZbL2KUx1SGWo3TocOb
TfB9K75T6PIGQ1AER/2jzH9+P7p6g1T7f0ySFnHIrOF9jboVmIup9DRd9LI64AmA
nPKBhxoeZdl2FckIQLG16R4yHqe8le+3cZPJb1dP3lFPjsKoOG4VGGew1VgtOm1D
40xRbtLBQj3PeQ2POMpVS4RReiXgWR544HIdR7sk4qteofH+53Cc85pDhJKN0ji+
CULT/WSrm5DixJCV2bw+
=+W/p
-----END PGP SIGNATURE-----
